utils/oscap-docker.8
.TH oscap-docker "8" "January 2016" "Red Hat, Inc." "System Administration Utilities"
.SH NAME
oscap-docker \- Tool for running oscap within docker container or image
.SH DESCRIPTION
oscap-docker tool can asses vulnerabilities or security compliance of running Docker
containers or cold Docker images. OpenSCAP tool \fBoscap(8)\fR is used underneath. Definition
of vulnerabilities (CVE stream) is downloaded from product vendor.

.SS Compliance scan of Docker image
Usage: oscap-docker image IMAGE_NAME OSCAP_ARGUMENT [OSCAP_ARGUMENT...]

Run any OpenSCAP \fBoscap(8)\fR command within chroot of mounted docker image. Learn more
about arguments in \fBoscap(8)\fR man page.

.SS Compliance scan of Docker container
Usage: oscap-docker container CONTAINER_NAME OSCAP_ARGUMENT [OSCAP_ARGUMENT...]

Run any OpenSCAP \fBoscap(8)\fR command within chroot of mounted docker container. Result
of this command may differ from scanning just an image due to defined mount points.

.SS "Vulnerability scan of Docker image"
Usage: oscap-docker image-cve IMAGE_NAME [--results oval-results-file.xml [--report report.html]]

Attach docker image, determine OS variant/version, download CVE stream applicable to
the given OS, and finally run vulnerability scan.

.SS "Vulnerability scap of Docker container"
Usage: oscap-docker container-cve CONTAINER_NAME [--results oval-results-file.xml [--report report.html]]

Chroot to running container, determine OS variant/version, download CVE stream applicable
to the given OS and finally run a vulnerability scan.

In order to use different \fBoscap(8)\fR binary pass it like --oscap=<path/to/oscap>, as the first argument.

.SH SECURITY POLICIES
.TP
\fB SCAP-Security-Guide\fR package contains multiple configuration policies.
.TP
\fB Red Hat CVE stream can be found online\fR - \fIhttps://www.redhat.com/security/data/metrics/\fR

.SH REPORTING BUGS
.nf
Please report bugs using https://github.com/OpenSCAP/openscap/issues

.SH AUTHORS
.nf
Šimon Lukašík <slukasik@redhat.com>
.fi


NEWS
openscap-1.3.1                                                  12-06-2019
  - New features
    - Support for SCAP 1.3 Source Datastreams (evaluating, XML schemas,
      validation)
    - Introduced `oscap-podman` -- a tool for SCAP evaluation of Podman
      images and containers (rhbz#1642373)
    - Tailoring files are included in ARF result files (#902)
    - OVAL details are always shown in HTML report, users do not have to
      provide `--oval-results` on command line
    - HTML report displays OVAL test details also for OVAL tests included
      from other OVAL definitions using `extend_definition` (#916, #954)
    - OVAL test IDs are shown in HTML report
    - Rule IDs are shown in HTML guide (#1293)
    - Added `block_size` in Linux `partition_state` defined in OVAL 5.11.2
    - Added `oscap_wrapper` that can be used to comfortably execute custom
      compiled oscap tool
  - Maintenance, bug fixes
    - Remote filesystems mounted using `autofs` direct maps are not
      recognized as local filesystems (rhbz#1655943)
    - SCAP source datastreams containing remote components can be
      evaluated without downloading remote data (rhbz#1709423)
    - Fixed duplicated variables in generated Ansible Playbooks
    - Fixed trailing whitespace characters in Ansible Playbooks
    - Correctly handle multiline profile titles and profile descriptions
      in generated Ansible Playbooks (#1112)
    - Fixed STIG Viewer output (--stig-viewer) to handle multiple rules
      that have the same STIG ID
    - Fixed incorrect displaying of OVAL test results in HTML report
    - Fixed segmentation fault in offline mode caused by usage of `chroot`
      file descriptor after closing (rhbz#1636431)
    - Fixed textfilecontent54 probe to not ignore `max_depth`, `recurse`,
      `recurse_direction` and `recurse_file_system` attributes of
      `behaviors` element when `filepath` element is given (rhbz#1655943)
    - Added CMake policies (CMP0078 and CMP0086) related to UseSWIG
    - Added RHEL 8 CPE, Fedora 31 CPE, Oracle Linux 8 CPE
    - Fedora CPEs fixed to work also on Fedora >= 30
    - Fixed segmentation fault in CVRF module (rhbz#1642283)
    - Fixed unresolved symbols in libopenscap_sce.so
    - Fixed memory leaks in Windows registry probe (#1269)
    - Fixed many GCC compiler warnings
    - Removed dead code from `fsdev` module
    - Many new test cases in upstream test suite
    - Refactoring
    - Updated Developer Guide
    - Updated manual pages

openscap-1.3.0                                                  09-10-2018
  - New features
    - Introduced a virtual '(all)' profile selecting all rules
    - Verbose mode is a global option in all modules
    - Added Microsoft Windows CPEs
    - oscap-ssh can supply SSH options into an environment variable
  - Maintenance
    - Removed SEXP parser
    - Added Fedora 30 CPE
    - Fixed many Coverity defects (memory leaks etc.)
    - SCE builds are enabled by default
    - Moved many low-level functions out of public API
    - Removed unused and dead code
    - Updated manual pages
    - Numerous small fixes

openscap-1.3.0_alpha2                                           10-08-2018
  - Maintenance
    - Removed '--probe-root' option
    - Removed '--show' option from 'oscap xccdf generate report'
    - Removed CCE API
    - Removed deprecated option '--sce-results'
    - Removed 'oscap oval list-probes' submodule
    - Removed 'validate-xml' submodule from CPE, OVAL, XCCDF modules
    - Moved OVAL probe handler to private headers
    - Added tests for filehash58 offline mode
    - Fixed broken SCE
    - Fixed problematic versioning in CMake and pkgconfig file
    - Removed many unused code
    - Rewritten test tests/API/XCCDF/default_cpe
    - Started to use asciidoc instead of asciidoctor
    - Fixed many compiler warnings
    - Fixed MinGW builds
    - Documentation updates
    - Small fixes

openscap-1.3.0_alpha1                                           18-07-2018
  - New features
    - Microsoft Windows support (issue #195)
    - new probes:
       - Windows registry probe
       - Windows accesstoken probe
       - Windows wmi57 probe
    - CMake is used as build system (issue #542)
    - CTest is used as test suite driver
  - Maintenance
    - probes are not separate processes, they are threads within oscap
    - OpenSCAP can be compiled using Visual Studio 2017
    - Dropped 53 deprecated API symbols (issue #1088)
    - Removed GNU Automake
    - Removed Python 2 support (issue #1034)
    - Ninja build is supported
    - Public API symbols are marked by OSCAP_API macro
    - Removed variable lenght arrays
    - Removed custom memory allocation functions (issue #1077)
    - Improved OS X build support
    - Fixed crash when deallocating red-black-tree node in Windows
    - Several large tests are splitted into smaller test cases
    - User manual is splitted in User and Developer manual
    - Many documentation updates (issue #1069, #1066)
    - Stopped using '\r' characters on stdout (issue #579, #1023)
    - Updated release tools to reflect CMake (issue #1036)
    - Dropped Cygwin support from User Manual (issue #1011)
    - source tarball does not contain build artifacts
    - Many small fixes

openscap-1.2.17                                                 29-05-2018
  - New features
    - HTML Guide user experience improvements
    - New options in HTML report "Group By" menu
    - oscap-ssh supports --oval-results (issue #863)
  - Maintenance
    - Support comparing state record elements with item
    - Updated Bash completion
    - Make Bash role headers consistent with --help output
    - Fixed problems reported by Coverity (issue #909)
    - Fixed CVE schema to support 4 to 7 digits CVEs
    - Fix output of generated bash role missing fix message
    - Fix oscap-docker to clean up temporary image (RHBZ #1454637)
    - Fix Ansible remediations generation
    - Add a newline between ids in xccdf info (issue #968)

utils/oscap-docker.py
''' oscap docker command '''
import argparse
from oscap_docker_python.oscap_docker_util import oscapScan
import docker
import sys
from requests import exceptions


class oscapDocker(object):
    ''' Generic class to call the scans '''
    def __init__(self):
        pass

    def set_args(self, args, unknown):
        '''
        Sets arguments for argparse into the oscapDocker class
        '''
        self.args = args
        self.unknown_args = unknown

    def cve_scan(self):
        ''' Wrapper function for container/image scanning '''
        OS = oscapScan()
        print OS.scan_cve(self.args.scan_target, self.unknown_args)

    def scan(self):
        ''' Wrapper functiopn to scan with openscap'''
        OS = oscapScan()
        print OS.scan(self.args.scan_target, self.unknown_args)


def ping_docker():
    ''' Simple check if the docker daemon is running '''
    client = docker.Client()
    client.ping()

if __name__ == '__main__':

    OD = oscapDocker()

    parser = argparse.ArgumentParser(description='oscap docker')
    subparser = parser.add_subparsers(help="commands")

    # Scan CVEs in image
    image_cve = subparser.add_parser('image-cve', help='Scan a docker image \
                                    for known vulnerabilities.')
    image_cve.set_defaults(func=OD.cve_scan)
    image_cve.add_argument('scan_target', help='Container or image to scan')

    # Scan an Image
    image = subparser.add_parser('image', help='Scan a docker image')
    image.add_argument('scan_target',
                       help='Container or image to scan')

    image.set_defaults(func=OD.scan)
    # Scan a container
    container = subparser.add_parser('container', help='Scan a running docker\
                                      container of given name.')
    container.add_argument('scan_target',
                           help='Container or image to scan')
    container.set_defaults(func=OD.scan)

    # Scan CVEs in container
    container_cve = subparser.add_parser('container-cve', help='Scan a \
                                         running container for known \
                                         vulnerabilities.')

    container_cve.set_defaults(func=OD.cve_scan)
    container_cve.add_argument('scan_target',
                               help='Container or image to scan')

    args, unknown = parser.parse_known_args()

    try:
        ping_docker()

    except exceptions.ConnectionError:
        print "The docker daemon does not appear to be running"
        sys.exit(1)

    OD.set_args(args, unknown)
    args.func()
